### ISSUE TITLE
add imagePullSecrets support

### ISSUE THREAD

**Issue by DazWilkin** [deislabs/krustlet#257]
I'm receiving this error from Krustlet trying to pull images from a public (!) GCR registry.

As shown below, `wasm-to-oci` can push|pull images to this registry and I've configured it to support public (non-authenticated) access. I'm able to run regular containers from it on the cluster without problem.

```
kubectl describe pod/hello-world-wasi-golang
Name:         hello-world-wasi-golang
Namespace:    default
Priority:     0
Node:         krustlet/
Labels:       <none>
Annotations:  Status:  Failed
Message:      Failed to decode registry token from auth request
IP:           
IPs:          <none>
Containers:
  hello-world-wasi-golang:
    Image:      gcr.io/${PROJECT}/hello-world-wasi-golang:v1
    Port:       <none>
    Host Port:  <none>
    Environment:
      FOO:             bar
      POD_NAME:        hello-world-wasi-golang (v1:metadata.name)
      CONFIG_MAP_VAL:  <set to the key 'myval' of config map 'hello-world-wasi-golang'>  Optional: false
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-x75t7 (ro)
Conditions:
  Type           Status
  PodScheduled   True 
Volumes:
  default-token-x75t7:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-x75t7
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     krustlet/arch=wasm32-wasi:NoExecute
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age        From               Message
  ----    ------     ----       ----               -------
  Normal  Scheduled  <unknown>  default-scheduler  Successfully assigned default/hello-world-wasi-golang to krustlet
```
I'm able to pull regular containers from the same registry to the same cluster.

```bash
PROJECT=....

docker pull hello-world
docker tag hello-world gcr.io/${PROJECT}/hello-world
docker push gcr.io/${PROJECT}/hello-world

kubectl run hello-world --image=gcr.io/${PROJECT}/hello-world

kubectl logs pod/hello-world

Hello from Docker!
...
```

I'm able to `wasm-to-oci [push|pull]` images to the registry:

```bash
go run ./cmd push \
  hello-world-wasi-golang \
  gcr.io/${PROJECT}/hello-world-wasi-golang:v1
INFO[0004] Pushed: gcr.io/${PROJECT}/hello-world-wasi-golang:v1 
INFO[0004] Size: 2289810                                
INFO[0004] Digest: sha256:0e047d23...

go run ./cmd pull gcr.io/${PROJECT}/hello-world-wasi-golang:v1
INFO[0001] Pulled: gcr.io/${PROJECT}/hello-world-wasi-golang:v1 
INFO[0001] Size: 2289810                                
INFO[0001] Digest: sha256:0e047d23...

```
I tried retrieving a copy of `hello-world-wasi-rust:v0.1.0` from `webassembly.azurecr.io`, adding it to `gcr.io/${PROJECT}/hello-world-wasi-rust:v1` and then reconfiguring the Pod YAML to reference it, but I get the same error.
```
go run ./cmd pull webassembly.azurecr.io/hello-world-wasi-rust:v0.1.0
INFO[0001] Pulled: webassembly.azurecr.io/hello-world-wasi-rust:v0.1.0 
INFO[0001] Size: 1964621                                
INFO[0001] Digest: sha256:670adc71...

go run ./cmd push module.wasm gcr.io/${PROJECT}/hello-world-wasi-rust:v1
INFO[0003] Pushed: gcr.io/${PROJECT}/hello-world-wasi-rust:v1 
INFO[0003] Size: 1964621                                
INFO[0003] Digest: sha256:54b09224...
```

**Comment by peterhuene** [comment]
This should be fixed by #255, but I haven't yet looked into the test failures there.

**Comment by DazWilkin** [comment]
If the GCR registry is configured to be public, @peterhuene proposed changes work for me:

```bash
go run ./cmd push \
./hello-world-wasi-rust \
gcr.io/${PROJECT}/hello-world-wasi-rust:v4
INFO[0001] Pushed: gcr.io/.../hello-world-wasi-rust:v4 
INFO[0001] Size: 1964621                                
INFO[0001] Digest: sha256:356ef906...
```
Then:
```bash
sed --in-place 's|hello-world-wasi-rust:v1|hello-world-wasi-rust:v4|g' rust.yaml
kubectl apply --filename=rust.yaml
kubectl describe pod/hello-world-wasi-rust
Name:         hello-world-wasi-rust
Namespace:    default
Priority:     0
Node:         krustlet/
Labels:       <none>
Annotations:  Status:  Succeeded
IP:           
IPs:          <none>
Containers:
  hello-world-wasi-rust:
    Container ID:  
    Image:         gcr.io/.../hello-world-wasi-rust:v4
```
and:
```
ls -l ${HOME}/.krustlet/.oci/modules/gcr.io/.../hello-world-wasi-rust
v1
v3
v4
```
If the GCR registry is configured to be private even if the cluster is configured to access the registry, I get a different error:

```bash
kubectl describe pod/hello-world-wasi-rust
Name:         hello-world-wasi-rust
Namespace:    default
Priority:     0
Node:         krustlet/
Labels:       <none>
Annotations:  Status:  Failed
Message:      error decoding response body: missing field `detail` at line 1 column 273
```

**Comment by peterhuene** [comment]
Pulling from a private registry requires supporting `imagePullSecrets` afaik.

My guess is that the response coming back is a 401/403 that's not JSON, perhaps needed accept headers aren't being set with the auth request?

**Comment by peterhuene** [comment]
I believe `imagePullSecrets` isn't supported yet for fetching images.

**Comment by bacongobbler** [comment]
Nope ^

**Comment by bacongobbler** [comment]
updated the title to reflect the current status.

**Comment by peterhuene** [comment]
The issue with authing against GCR was fixed in #255.

**Comment by itowlson** [comment]
I've got a proof of concept that works with ACR (though I'm not sure if I'm doing the right thing for the general case!).  Will pick this up and refine my solution into something more product-like unless anybody else wants to claim it.
